[dist_optim] introduce distributed functional optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45221 [dist_optim] introduce distributed functional optimizer**
* #44791 [optimizer] refactor Adam to use functional API
* #44715 [optimizer] introduce optimizer functional API, refactor Adagrad

This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

Differential Revision: [D23935256](https://our.internmc.facebook.com/intern/diff/D23935256)

wanchaol committed:
[WIP][dist_optim] introduce distributed functional optimizer

This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

pritamdamania87 commented on torch/optim/functional.py:
```diff
@@ -6,6 +6,13 @@
 
 # TODO: use foreach API in optim.functional to do all the computation
 
+def _make_sparse(grad, grad_indices, values):
```
@vincentqb Could you review these changes?

vincentqb replied:
LGTM. What is the advantage of writing this in this way?

wanchaol replied:
It's mainly because nested functions are not supported by TorchScript, so I moved it out as a separate function. Also I feel like define a nested function inside a for loop is not a good practice

pritamdamania87 commented on torch/distributed/optim/optimizer.py:
```diff
@@ -137,10 +188,10 @@ def step(self, context_id):
             rpc_futs.append(rpc.rpc_async(
-                optim.owner(),
+                optimizer.owner(),
                 _local_optimizer_step,
```
Can we have two versions of `_local_optimizer_step`? If we're using a torchscript functional optimizer, we can have a torchscript `_local_optimizer_step` which means RPC will call an torchscript function and there would be no GIL anywhere. If it is an optimizer that is not torchscript we call a regular python `_local_optimizer_step`.

In the current version, we still end up holding GIL since `_local_optimizer_step` is not torchscript.

wanchaol replied:
hmm I see, this make sense, the problem of this is that `_LocalOptimizer` still contains `Lock()` objects to protect those who doesn't have a functional optimizer implementation. I could try adding a new `_LocalScriptOptimizer` to enable the use case you mentioned

pritamdamania87 reviewed (changes requested):
Looks good overall, requesting changes to see if we can further optimize the local optimizer step to be completely in torchscript.

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] introduce distributed functional optimizer"


This PR introduces a distributed functional optimizer, so that
distributed optimizer can reuse the functional optimizer APIs and
maintain their own states. This could enable the torchscript compatible
functional optimizer when using distributed optimizer, helps getting rid
of GIL and improve overall performance of training, especially distributed
model parallel training

[ghstack-poisoned]
